src/controller/CommissioningWindowOpener.cpp

```diff
@@ -0,0 +1,348 @@
+    mCommissioningWindowIteration     = iteration;
+
+    bool randomSetupPIN = !setupPIN.HasValue();
+    ReturnErrorOnFailure(PASESession::GeneratePASEVerifier(mVerifier, mCommissioningWindowIteration, GetSPAKE2Salt(),
+                                                           randomSetupPIN, mSetupPayload.setUpPINCode));
+
+    payload = mSetupPayload;
+
+    if (readVIDPIDAttributes)
+    {
+        mNextStep = Step::kReadVID;
+    }
+    else
+    {
+        mNextStep = Step::kOpenCommissioningWindow;
+    }
+
+    return mController->GetConnectedDevice(mNodeId, &mDeviceConnected, &mDeviceConnectionFailure);
+}
+
+CHIP_ERROR CommissioningWindowOpener::OpenCommissioningWindowInternal(OperationalDeviceProxy * device)
+{
+    ChipLogProgress(Controller, "OpenCommissioningWindow for device ID %" PRIu64, mNodeId);
+
+    constexpr EndpointId kAdministratorCommissioningClusterEndpoint = 0;
+
+    AdministratorCommissioningCluster cluster;
+    cluster.Associate(device, kAdministratorCommissioningClusterEndpoint);
+
+    if (mCommissioningWindowOption != CommissioningWindowOption::kOriginalSetupCode)
+    {
+        chip::Spake2pVerifierSerialized serializedVerifier;
+        MutableByteSpan serializedVerifierSpan(serializedVerifier);
+        ReturnErrorOnFailure(mVerifier.Serialize(serializedVerifierSpan));
+
+        AdministratorCommissioning::Commands::OpenCommissioningWindow::Type request;
+        request.commissioningTimeout = mCommissioningWindowTimeout;
+        request.PAKEVerifier         = serializedVerifierSpan;
+        request.discriminator        = mSetupPayload.discriminator;
+        request.iterations           = mCommissioningWindowIteration;
+        request.salt                 = GetSPAKE2Salt();
+
+        // TODO: What should the timed invoke timeout here be?
+        uint16_t timedInvokeTimeoutMs = 10000;
+        ReturnErrorOnFailure(cluster.InvokeCommand(request, this, OnOpenCommissioningWindowSuccess,
+                                                   OnOpenCommissioningWindowFailure, MakeOptional(timedInvokeTimeoutMs)));
+
+        char payloadBuffer[QRCodeBasicSetupPayloadGenerator::kMaxQRCodeBase38RepresentationLength];
+
+        MutableCharSpan manualCode(payloadBuffer);
+        ReturnErrorOnFailure(ManualSetupPayloadGenerator(mSetupPayload).payloadDecimalStringRepresentation(manualCode));
+        ChipLogProgress(Controller, "Manual pairing code: [%s]", payloadBuffer);
+
+        MutableCharSpan QRCode(payloadBuffer);
+        ReturnErrorOnFailure(QRCodeBasicSetupPayloadGenerator(mSetupPayload).payloadBase38Representation(QRCode));
+        ChipLogProgress(Controller, "SetupQRCode: [%s]", payloadBuffer);
+    }
+    else
+    {
+        AdministratorCommissioning::Commands::OpenBasicCommissioningWindow::Type request;
+        request.commissioningTimeout = mCommissioningWindowTimeout;
+        // TODO: What should the timed invoke timeout here be?
+        uint16_t timedInvokeTimeoutMs = 10000;
```

Comment:
these could share a `constexpr` maintained at the top of the file together with the todo as they seem the same as on line 120.

Reply:
Pre-existing, but done.